### PR TITLE
Mediation Policy Provider

### PR DESCRIPTION
**Pull request by ThusharaSampath**
This PR introduces the following features

- An abstraction layer for mediation policy operations
- A service to provide default functions of mediation
- Policy versioning support
- An Internal API to check whether mediations are attached to an API
- fixes for choreo-organization support in various flows

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/UserTokenUtil.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/UserTokenUtil.java:2]: LLC

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ApisApiServiceImpl.java [hunk 2]
```diff
@@ -143,4 +146,36 @@ public Response unDeployedAPIRevision(UnDeployedAPIRevisionDTO unDeployedAPIRevi
                 unDeployedAPIRevisionDTO.getEnvironment());
         return Response.ok().build();
     }
+
+    @Override
+    public Response apisApiIdOperationPolicyIsAttachedGet(String xWSO2Tenant, String apiId, String organizationId,
+                                                          MessageContext messageContext) throws APIManagementException {
+        try {
+            String organization;
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ApisApiServiceImpl.java:154]: `String organization = organizationId == null ? SubscriptionValidationDataUtil.validateTenantDomain(xWSO2Tenant, messageContext) : organizationId;`

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ApisApiServiceImpl.java [hunk 3]
```diff
@@ -143,4 +146,36 @@ public Response unDeployedAPIRevision(UnDeployedAPIRevisionDTO unDeployedAPIRevi
...
+            OperationPolicyAttachmentStatusDTO result = new OperationPolicyAttachmentStatusDTO();
+            result.isAttached(isAttached);
+            return Response.ok().entity(result).build();
+
+        } catch (APIManagementException e) {
+            RestApiUtil.handleBadRequest(
+                    "An error has occurred while getting API for the given organization/tenantDomain and apiId combination.",
+                    e, log);
+        }
+        return Response.ok().build();
```
Comment (menakaj) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ApisApiServiceImpl.java:179]: If policies are not attached, let's return `{isAttached:false}`

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/OperationPolicyProvider.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/OperationPolicyProvider.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/OperationPolicyProvider.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+     * Get common operation policy for given name
+     *
+     * @param name         policyName
+     * @param version      PolicyVersion
+     * @param organization organization
+     * @return Operation policy data
+     * @throws APIManagementException failed to get common operation policy
+     */
+    OperationPolicyData getCommonOperationPolicyByPolicyName
+    (String name, String version, String organization, boolean isWithPolicyDefinition) throws APIManagementException;
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/OperationPolicyProvider.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 6]
```diff
@@ -369,6 +370,19 @@ private void readChildElements(OMElement serverConfig,
...
                 persistenceProperties = persistenceProps;
+            } else if (APIConstants.ApiManagerConfigurationConstants.OPERATION_POLICY_CONFIGS.equals(localName)) {
+                OMElement properties = element
+                        .getFirstChildWithName(new QName(APIConstants.ApiManagerConfigurationConstants.PROPERTIES));
+                Iterator policyProviderPropertiesIterator = properties
+                        .getChildrenWithLocalName(APIConstants.ApiManagerConfigurationConstants.PROPERTY);
+                Map<String, String> operationPolicyProps = new HashMap<>();
+                while (policyProviderPropertiesIterator.hasNext()) {
+                    OMElement propertyElem = (OMElement) policyProviderPropertiesIterator.next();
+                    String name = propertyElem.getAttributeValue(new QName(APIConstants.ApiManagerConfigurationConstants.NAME));
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 7]
```diff
@@ -1877,10 +1876,8 @@ private void cloneOperationPoliciesToAPI(String oldAPIUuid, API newAPI,
                 for (OperationPolicy operationPolicy : operationPolicies) {
                     String clonedPolicyId;
                     if (!clonedPolicies.containsKey(operationPolicy.getPolicyId())) {
-                        OperationPolicyData apiSpecificOperationPolicy =
-                                apiMgtDAO.getAPISpecificOperationPolicyByPolicyID(operationPolicy.getPolicyId(),
-                                        oldAPIUuid, newAPI.getOrganization(), true);
-                        clonedPolicyId = apiMgtDAO.cloneOperationPolicy(newAPI.getUuid(), apiSpecificOperationPolicy);
+                        clonedPolicyId = operationPolicyProviderInstance.cloneAPISpecificOperationPolicy(operationPolicy.getPolicyId(), oldAPIUuid,
+                                newAPI.getUuid(), newAPI.getOrganization());
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderImpl.java [hunk 8]
```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderImpl.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderImpl.java [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+import org.wso2.carbon.apimgt.api.APIManagementException;
+import org.wso2.carbon.apimgt.api.OperationPolicyProvider;
+import org.wso2.carbon.apimgt.api.model.OperationPolicyData;
+import org.wso2.carbon.apimgt.api.model.OperationPolicySpecification;
+import org.wso2.carbon.apimgt.impl.dao.ApiMgtDAO;
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+
+import java.util.List;
+
+public class OperationPolicyProviderImpl implements OperationPolicyProvider {
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderImpl.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderImpl.java [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+     *
+     * @param policyId UUID of the policy to be deleted
+     * @throws APIManagementException
+     */
+    @Override
+    public void deleteOperationPolicyById(String policyId, String organization) throws APIManagementException {
+        apiMgtDAO.deleteOperationPolicyByPolicyId(policyId);
+    }
+
+    // API SPECIFIC
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderImpl.java]: No.. fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderImpl.java [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+    public void cloneAllApiSpecificOperationPolicies(String oldApiUUID, String newApiUUID, String organization)
+            throws APIManagementException {
+        List<OperationPolicyData> operationPolicyDataList = getAllAPiSpecificOperationPolicies(oldApiUUID, organization);
+        for (OperationPolicyData policyData : operationPolicyDataList) {
+            apiMgtDAO.cloneOperationPolicy(newApiUUID, policyData);
+        }
+
+    }
+
+    // Other
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderImpl.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderManager.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/OperationPolicyProviderManager.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/UserTokenUtil.java (hunk 1 above)
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/UserTokenUtil.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ApisApiServiceImpl.java (hunk 2 above)
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ApisApiServiceImpl.java]: fixed with https://github.com/wso2/carbon-apimgt/pull/11600/commits/db32ac912f17e8a7e3f2cf633a0955376d6b2a2a

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ApisApiServiceImpl.java (hunk 3 above)
Comment (ThusharaSampath) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ApisApiServiceImpl.java:175]: currently, it does return `{isAttached:false}`

**Pull request merged**